File: 23-merge-k-sorted-lists/merge-k-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeKLists(struct ListNode** lists, int listsSize) {
    struct ListNode * head = NULL;
    int head_index;
    for (int i = 0; i < listsSize; i++) {
        if ((lists[i] != NULL && head == NULL) ||
            (lists[i] != NULL && head != NULL && head->val > lists[i]->val)) {
            head = lists[i];
            head_index = i;
        }
    }

    if (head == NULL) {
        return NULL;
    }

    lists[head_index] = lists[head_index]->next;
    head->next = mergeKLists(lists, listsSize);
    return head;
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        min_heap = []
    
        for i, lst in enumerate(lists):
            if lst:
                heapq.heappush(min_heap, (lst.val, i, lst))
        
        dummy = ListNode()
        current = dummy
        
        while min_heap:
            val, i, node = heapq.heappop(min_heap)
            
            current.next = node
            current = current.next
            
            if node.next:
                heapq.heappush(min_heap, (node.next.val, i, node.next))
        
        return dummy.next